extensions/amp-story/1.0/amp-story.js

```diff
@@ -1284,17 +1339,57 @@ export class AmpStory extends AMP.BaseElement {
       return Promise.resolve();
     }
 
-    if (
-      isExperimentOn(this.win, 'amp-story-paywall-exp') &&
-      targetPage.isPaywallProtected()
-    ) {
-      if (this.storeService_.get(StateProperty.SUBSCRIPTIONS_DIALOG_STATE)) {
-        // Subscription dialog is already triggered.
-        return Promise.resolve();
-      }
-      this.storeService_.dispatch(Action.TOGGLE_SUBSCRIPTIONS_DIALOG, true);
+    const subscriptionsState = this.storeService_.get(
+      StateProperty.SUBSCRIPTIONS_STATE
+    );
+    if (this.isPaywallStory_()) {
+      if (
+        pageIndex >= PAYWALL_PAGE_INDEX &&
+        subscriptionsState !== SubscriptionsState.GRANTED
+      ) {
+        // Block while waiting for upcoming entitlement.
+        if (subscriptionsState === SubscriptionsState.UNKNOWN) {
+          return this.subscriptionsStatePromise_.promise.then(() => {
+            return this.switchTo_(targetPageId, direction);
+          });
+        }
+        if (!this.pageAfterGranted_) {
+          this.pageAfterGranted_ = targetPageId;
+        }
+
+        // Attempt to navigate to the locked pages after the paywall page(the first blocked page) with blocked state.
+        if (pageIndex > PAYWALL_PAGE_INDEX) {
+          // Show the paywall immediately if the active page is the paywall page.
+          if (this.getPageIndex(this.activePage_) === PAYWALL_PAGE_INDEX) {
```

Comment:
how can this ever be true if prev line is `>`

Reply:
When the current index is 2 and the target index is 3(e.g. tapping right on paywall page), the target page index is larger than `PAYWALL_PAGE_INDEX` and the current index equals to `PAYWALL_PAGE_INDEX`.